core/core.go

```diff
@@ -557,6 +630,103 @@ func (t *Textile) noAuthFunc(ctx context.Context) (context.Context, error) {
 	return ctx, nil
 }
 
+func generatePowUnaryInterceptor(serviceName string, allowedMethods []string, serviceDesc *desc.ServiceDescriptor, stub *grpcdynamic.Stub, pc *powc.Client, c *mdb.Collections) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		methodParts := strings.Split(info.FullMethod, "/")
+		if len(methodParts) != 3 {
+			return nil, status.Errorf(codes.Internal, "error parsing method string %s", info.FullMethod)
+		}
+		methodServiceName := methodParts[1]
+		methodName := methodParts[2]
+
+		if methodServiceName != serviceName {
+			return handler(ctx, req)
+		}
+
+		if pc == nil {
+			return nil, status.Error(codes.Internal, "Powergate isn't enabled in Hub")
+		}
+
+		isAllowedMethod := false
+		for _, method := range allowedMethods {
+			if method == methodName {
+				isAllowedMethod = true
+				break
+			}
+		}
+
+		if !isAllowedMethod {
+			return nil, status.Errorf(codes.PermissionDenied, "method not allowed: %s", info.FullMethod)
+		}
+
+		var ffsInfo *mdb.FFSInfo
+		var owner crypto.PubKey
+		var isAccount bool
+		if org, ok := mdb.OrgFromContext(ctx); ok {
+			ffsInfo = org.FFSInfo
+			owner = org.Key
+			isAccount = true
+		} else if dev, ok := mdb.DevFromContext(ctx); ok {
+			ffsInfo = dev.FFSInfo
+			owner = dev.Key
+			isAccount = true
+		} else if user, ok := mdb.UserFromContext(ctx); ok {
+			ffsInfo = user.FFSInfo
+			owner = user.Key
+			isAccount = false
+		}
+
+		createNewFFS := func() error {
+			id, token, err := pc.FFS.Create(ctx)
+			if err != nil {
+				return fmt.Errorf("creating new ffs instance: %v", err)
+			}
+			if isAccount {
+				_, err = c.Accounts.UpdateFFSInfo(ctx, owner, &mdb.FFSInfo{ID: id, Token: token})
+			} else {
+				_, err = c.Users.UpdateFFSInfo(ctx, owner, &mdb.FFSInfo{ID: id, Token: token})
+			}
+			if err != nil {
+				return fmt.Errorf("updating user/account with new ffs information: %v", err)
+			}
+			return nil
+		}
+
+		tryAgain := fmt.Errorf("powergate newly integrated into your account, please try again in 30 seconds to allow time for setup to complete")
+
+		// case where account/user was created before powergate was enabled.
+		// create a ffs instance for them.
+		if ffsInfo == nil {
+			if err := createNewFFS(); err != nil {
+				return nil, err
+			}
+			return nil, tryAgain
```

Comment:
Even though not all calls to powergate rely on some time passing for wallet funding to complete, it's just easy/safe to return this error instead of building recursive retry logic.